torch/_torch_docs.py

```diff
@@ -3385,6 +3385,55 @@ def merge_dicts(*dicts):
     array([-1,  2,  3])
 """)
 
+add_docstr(torch.from_buffer,
+           r"""
+from_buffer(buffer, dtype=None, count=-1, offset=0, *, device=None, requires_grad=False) -> Tensor
+
+Create a 1-dimensional :class:`Tensor` from an object that implements
```

Comment:
I was thinking about this more and wondering how the Python Array API deals with the buffer protocol. Turns out it has an `asarray` function:

https://numpy.org/doc/stable/reference/generated/numpy.frombuffer.html?highlight=frombuffer#numpy.frombuffer

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

Since our new focus is on implementing the Python Array API, do you think PyTorch would want both `asarray` and `frombuffer`, @ysiraichi, or should we just focus on building the `asarray` function? This work would be required to do it, and I'm curious to hear your thoughts.

Reply:
Another option is to implement a `frombuffer` function (name TBD) that was compliant with `asarray`, so we could then build `asarray` using it

Reply:
I think that's effectively why `numpy.frombuffer` is little used and doesn't quite make sense anymore - `numpy.asarray` does the same and more, and is clearer. The (PEP 3118) buffer protocol is C only and not supposed to be user-visible. Both `numpy.frombuffer` and PEP 3118 are ~15 years old by now, so I can't remember or easily find the history there anymore.

Reply:
I think it makes sense to have both, while calling `frombuffer` in `asarray`. If we are to follow the Python Array API, I guess `frombuffer` still enables users to specify an offset and count (which is nice, but the frequency they are used probably wouldn't be high).

Reply:
That sounds reasonable -- one question about that, though: is `asarray`'s buffer behavior the same as `frombuffer`'s?

Reply:
Given that the input object is something that implements the buffer protocol, they are the same if you consider `offset=0` and `count=-1` arguments for `frombuffer`. From what I could understand, `asarray` is a function that "does the right thing", depending on what the input object is (almost a superset of `frombuffer`).

Actually, I would say that `asarray` is somewhat analogous to what we had with the legacy tensor constructor. It kind of tried to do the right thing in that context. So, in that sense, maybe we would rather have `frombuffer`, `fromdlpack`... The good thing about `asarray`, in my opinion, is that it's standardized in the Python Array API.

> `numpy.asarray` does the same and more, and is clearer.

I think `numpy.asarray` is a bit different than what the Python Array API specifies. I think it is closer to PyTorch's `as_tensor` (without the `order` parameter).

https://pytorch.org/docs/stable/generated/torch.as_tensor.html

Reply:
Interesting! As long as this version of `frombuffer` is consistent from NumPy's `frombuffer` AND can be used as a helper by a future `asarray` (as defined by the Python Array API) I think it's OK.

Reply:
I believe, as is, it's consistent with NumPy's version. It can also easily be used by a future `asarray` function, by just hoisting the code in that function into a helper one (I guess we can leave it for that PR). The only thing now is the name of the function. There are consistency reasons for both `frombuffer` and `from_buffer`, in my opinion.
- `frombuffer` is consistent with NumPy's version
- `from_buffer` is consistent with PyTorch's `Storage::from_buffer` and `torch.from_numpy`
What do you think?

Reply:
Great question. I prefer `frombuffer` myself for consistency with NumPy. Sound good?

Reply:
Yes. I will make the change!